onadata/apps/logger/models/instance.py

```diff
@@ -203,6 +209,71 @@ def _set_uuid(self):
                 self.uuid = uuid
         set_uuid(self)
 
+    def _populate_xml_hash(self):
+        '''
+        Populate the `xml_hash` attribute of this `Instance` based on the content of the `xml`
+        attribute.
+        '''
+        self.xml_hash = self.get_hash(self.xml)
+
+    @classmethod
+    def populate_xml_hashes_for_instances(cls, usernames=None, pk__in=None, repopulate=False):
+        '''
+        Populate the `xml_hash` field for `Instance` instances limited to the specified users
+        and/or DB primary keys.
+
+        :param list[str] usernames: Optional list of usernames for whom `Instance`s will be
+        populated with hashes.
+        :param list[int] pk__in: Optional list of primary keys for `Instance`s that should be
+        populated with hashes.
+        :param bool repopulate: Optional argument to force repopulation of existing hashes.
+        :returns: Total number of `Instance`s updated.
+        :rtype: int
+        '''
+
+        filter_kwargs = dict()
+        if usernames:
+            # Convert usernames to `User` objects.
+            user__in = User.objects.filter(username__in=usernames)
+            filter_kwargs.update({'user__in': user__in})
+        if pk__in:
+            filter_kwargs.update({'pk__in': pk__in})
+        # By default, skip over instances previously populated with hashes.
+        if not repopulate:
+            filter_kwargs.update({'xml_hash': cls.DEFAULT_XML_HASH})
+
+        # Query for the target `Instance`s.
+        target_instances_queryset = cls.objects.filter(**filter_kwargs)
+
+        # Exit quickly if there's nothing to do.
+        if not target_instances_queryset.exists():
+            return 0
+
+        # Limit our queryset result content since we'll only need the `pk` and `xml` attributes.
+        target_instances_queryset = target_instances_queryset.only('pk', 'xml')
+        instances_updated_total = 0
+
+        # Break the potentially large `target_instances_queryset` into chunks to avoid memory
+        # exhaustion.
+        chunk_size = 2000
+        while target_instances_queryset.exists():
```

Comment:
This loops forever if `repopulate=True`, since the query set in that case is not limited to instances whose `xml_hash=cls.DEFAULT_XML_HASH`